File: yggdrasil_decision_forests/utils/hyper_parameters.cc
/*
 * Copyright 2021 Google LLC.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "yggdrasil_decision_forests/utils/hyper_parameters.h"

#include <string>
#include <utility>

#include "absl/container/flat_hash_map.h"
#include "absl/container/flat_hash_set.h"
#include "absl/meta/type_traits.h"
#include "absl/status/status.h"
#include "absl/strings/str_cat.h"
#include "absl/strings/string_view.h"
#include "yggdrasil_decision_forests/learner/abstract_learner.pb.h"
#include "yggdrasil_decision_forests/utils/logging.h"

namespace yggdrasil_decision_forests {
namespace utils {

GenericHyperParameterConsumer::GenericHyperParameterConsumer(
    const model::proto::GenericHyperParameters& generic_hyper_parameters) {
  for (const auto& field : generic_hyper_parameters.fields()) {
    if (generic_hyper_parameters_.find(field.name()) !=
        generic_hyper_parameters_.end()) {
      LOG(FATAL) << "The field \"" << field.name()
                 << "\" is defined several times.";
    }
    generic_hyper_parameters_[field.name()] = field;
  }
}

absl::optional<model::proto::GenericHyperParameters::Field>
GenericHyperParameterConsumer::Get(const absl::string_view key) {
  if (consumed_values_.find(key) != consumed_values_.end()) {
    LOG(FATAL) << absl::StrCat("Already consumed hyper-parameter \"", key,
                               "\".");
  }
  consumed_values_.insert(std::string(key));
  const auto value_it = generic_hyper_parameters_.find(key);
  if (value_it == generic_hyper_parameters_.end()) {
    return {};
  }
  return value_it->second;
}

absl::Status
GenericHyperParameterConsumer::CheckThatAllHyperparametersAreConsumed() const {
  for (const auto& field : generic_hyper_parameters_) {
    if (consumed_values_.find(field.first) == consumed_values_.end()) {
      return absl::InvalidArgumentError(
          absl::StrCat("Unused hyper-parameter \"", field.first, "\"."));
    }
  }
  return absl::OkStatus();
}

utils::StatusOr<bool> SatisfyDefaultCondition(
    const model::proto::GenericHyperParameterSpecification::Value& value,
    const model::proto::GenericHyperParameterSpecification::Conditional&
        condition) {
  switch (condition.constraint_case()) {
    case model::proto::GenericHyperParameterSpecification::Conditional::
        kCategorical:
      if (!value.has_categorical()) {
        return absl::InvalidArgumentError("The value is not categorical.");
      }
      return std::find(condition.categorical().values().begin(),
                       condition.categorical().values().end(),
                       value.categorical().default_value()) !=
             condition.categorical().values().end();
    default:
      return absl::InvalidArgumentError("Invalid condition");
  }
}

bool HyperParameterIsBoolean(
    const model::proto::GenericHyperParameterSpecification::Value& def) {
  return def.has_categorical() &&
         def.categorical().possible_values_size() == 2 &&
         ((def.categorical().possible_values(0) == "false" &&
           def.categorical().possible_values(1) == "true") ||
          (def.categorical().possible_values(0) == "true" &&
           def.categorical().possible_values(1) == "false"));
}

}  // namespace utils
}  // namespace yggdrasil_decision_forests

File: yggdrasil_decision_forests/port/python/ydf/learner/hyperparameters.py
# Copyright 2022 Google LLC.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Utility functions for YDF Hyperparameters."""

from collections.abc import Mapping
import dataclasses
from typing import Dict, Optional, Set, Union

from yggdrasil_decision_forests.learner import abstract_learner_pb2
from yggdrasil_decision_forests.model import hyperparameter_pb2
from ydf.cc import ydf
from ydf.learner import custom_loss


HyperParameters = Dict[
    str, Optional[Union[int, float, str, bool, custom_loss.AbstractCustomLoss]]
]


def dict_to_generic_hyperparameter(
    src: HyperParameters,
) -> hyperparameter_pb2.GenericHyperParameters:
  """Transform a dictionary of hyperparameters to the corresponding proto."""
  generic_hps = hyperparameter_pb2.GenericHyperParameters()
  for key, value in src.items():
    if value is None:
      # The value is not defined, use default.
      continue
    if key == "loss" and isinstance(value, custom_loss.AbstractCustomLoss):
      # Custom Python fields must be treated separately.
      continue
    # Boolean has to come first, since it is a subtype of int.
    if isinstance(value, bool):
      # The GenericHyperParameters proto demands this conversion.
      value_as_str = "true" if value else "false"
      generic_hps.fields.append(
          hyperparameter_pb2.GenericHyperParameters.Field(
              name=key,
              value=hyperparameter_pb2.GenericHyperParameters.Value(
                  categorical=value_as_str
              ),
          )
      )
    elif isinstance(value, int):
      generic_hps.fields.append(
          hyperparameter_pb2.GenericHyperParameters.Field(
              name=key,
              value=hyperparameter_pb2.GenericHyperParameters.Value(
                  integer=value
              ),
          )
      )
    elif isinstance(value, float):
      generic_hps.fields.append(
          hyperparameter_pb2.GenericHyperParameters.Field(
              name=key,
              value=hyperparameter_pb2.GenericHyperParameters.Value(real=value),
          )
      )
    elif isinstance(value, str):
      generic_hps.fields.append(
          hyperparameter_pb2.GenericHyperParameters.Field(
              name=key,
              value=hyperparameter_pb2.GenericHyperParameters.Value(
                  categorical=value
              ),
          )
      )
    else:
      raise ValueError(f"Invalid value {value} for parameter {key}")
  return generic_hps


def validate_hyperparameters(
    hp_dict: HyperParameters,
    train_config: abstract_learner_pb2.TrainingConfig,
    deployment_config: abstract_learner_pb2.DeploymentConfig,
):
  """Returns None if the hyperparameters are valid, raises otherwise."""
  not_none_hps = set(key for key, value in hp_dict.items() if value is not None)
  return ydf.ValidateHyperparameters(
      not_none_hps, train_config, deployment_config
  )


def fix_hyperparameters(
    hp_dict: HyperParameters,
    explicit_parameters: Set[str],
    train_config: abstract_learner_pb2.TrainingConfig,
    deployment_config: abstract_learner_pb2.DeploymentConfig,
) -> HyperParameters:
  """Returns exclusion-free hyperparameters."""
  not_none_hp_names = set(
      key for key, value in hp_dict.items() if value is not None
  )
  explicit_hp_names = explicit_parameters.intersection(not_none_hp_names)
  invalid_hp_names = ydf.GetInvalidHyperparameters(
      not_none_hp_names, explicit_hp_names, train_config, deployment_config
  )
  return {
      key: value
      for key, value in hp_dict.items()
      if key not in invalid_hp_names
  }


@dataclasses.dataclass
class HyperparameterTemplate(Mapping):
  """A named and versioned set of hyper-parameters.

  List of hyper-parameter sets that outperforms the default hyper-parameters
  (either generally or in specific scenarios). A template is also a mapping of
  hyperparameters and may be used with the double star operator.

  Usage example:

  ```python
  templates = ydf.GradientBoostedTreesLearner.hyperparameter_templates()
  better_default = templates["better_defaultv1"]
  # Apply the parameters of the template on the learner.
  learner = ydf.GradientBoostedTreesLearner(label, **better_default)
  ```
  """

  name: str
  version: int
  parameters: HyperParameters
  description: str

  def __iter__(self):
    for key in self.parameters.keys():
      yield key

  def __len__(self):
    return len(self.parameters)

  def __getitem__(self, item):
    if isinstance(self.parameters, dict) and item in self.parameters:
      return self.parameters[item]
    return None
